### ISSUE TITLE
Can't alter the learning rate though loading optimizer states

### ISSUE THREAD

**Issue by Alex17Li** [pytorch/pytorch#156846]
### 🐛 Describe the bug

When loading an optimizer state for finetuning, the optimizer lr is scaled to follow the new scheduler with the starting lr that it was at last time instead of the specified starting lr. Unless I am doing something wrong

```python
import torch

from torch.optim import Adam

def create_optimizer(lr, model):
    return Adam([{"params": filter(lambda p: p.requires_grad, model.parameters()), "lr": lr}],  lr)

class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.conv1 = torch.nn.Conv2d(1, 2, 3)
        self.conv2 = torch.nn.Conv2d(1, 2, 4)
        self.new_parameters = self.conv1.parameters()
        self.pretrained_parameters = self.conv2.parameters()


def test_optimizer_scheduler_linear_from_opt_checkpoint_torchonly() -> None:
    lr = 10.0
    expected_order = [0.8, 0.6, 0.4, 0.2, 0]
    opt = create_optimizer(lr, Model())
    scheduler = torch.optim.lr_scheduler.LinearLR(
        opt,
        start_factor=0.8,
        end_factor=0.0,
        total_iters=4,
    )
    
    # Run some steps
    for i in range(1):
        print(f'Got {opt.param_groups[0]["lr"]} expected {lr * expected_order[i]}')
        opt.step()
        scheduler.step()
        print(f'Got {opt.param_groups[0]["lr"]} expected {lr * expected_order[i + 1]}')

    print("Reload")
    # Simulate new training run, this time we reload optimizer checkpoint only (e.g. fine tuning)
    state = {
        "opt": opt.state_dict(),
    }
    new_lr = 100.0 # optimizer outputs don't change regardless of this variable. 
    opt = create_optimizer(new_lr, Model())
    scheduler = torch.optim.lr_scheduler.LinearLR(
        opt,
        start_factor=0.8,
        end_factor=0.0,
        total_iters=4,
    )
    opt.load_state_dict(state_dict=state['opt']) # Note: You must load opt state dict after creating the scheduler or the lr will be wrong in the case where you load both the opt and scheduler (load must happen at the same time).

    for i in range(len(expected_order) - 1):
        print(f'Got {opt.param_groups[0]["lr"]} expected {new_lr * expected_order[i]}')
        opt.step()
        scheduler.step()


if __name__ == "__main__":
    print(f"{torch.__version__=}")

    test_optimizer_scheduler_linear_from_opt_checkpoint_torchonly() 
```

```
torch.__version__='2.7.1+cu126'
Got 8.0 expected 8.0
Got 6.0 expected 6.0
Reload
Got 6.0 expected 80.0
Got 4.5 expected 60.0
Got 3.0000000000000004 expected 40.0
Got 1.5000000000000002 expected 20.0
```

### Versions

```
Collecting environment information...
PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.28.1
Libc version: glibc-2.35

Python version: 3.10.0 (default, Mar  3 2022, 09:58:08) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-6.8.0-51-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4070 Ti
Nvidia driver version: 550.144.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               24
On-line CPU(s) list:                  0-23
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i7-13700K
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          5400.0000
CPU min MHz:                          800.0000
BogoMIPS:                             6835.20
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            640 KiB (16 instances)
L1i cache:                            768 KiB (16 instances)
L2 cache:                             24 MiB (10 instances)
L3 cache:                             30 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-23
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.6.3
[pip3] flake8==3.9.2
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.1
[pip3] numpy-quaternion==2022.4.3
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.17.0
[pip3] onnx-graphsurgeon==0.3.27
[pip3] onnxruntime==1.17.1
[pip3] onnxsim==0.4.36
[pip3] pie-torch==0.0.1
[pip3] pytorch-lightning==1.1.8
[pip3] segmentation-models-pytorch==0.1.3
[pip3] torch==2.7.1
[pip3] torchaudio==0.12.1+cu116
[pip3] torchmetrics==1.3.2
[pip3] torchvision==0.17.2
[pip3] torchviz==0.0.2
[pip3] triton==3.3.1
[conda] efficientnet-pytorch      0.6.3                    pypi_0    pypi
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] numpy-quaternion          2022.4.3                 pypi_0    pypi
[conda] nvidia-cublas-cu11        11.10.3.66               pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu11    11.7.101                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu11    11.7.99                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu11  11.7.99                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu11         8.5.0.96                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu11         10.9.0.58                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu11        10.2.10.91               pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu11      11.4.0.1                 pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu11      11.7.4.91                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu11          2.14.3                   pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu11          11.7.91                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] pie-torch                 0.0.1                    pypi_0    pypi
[conda] pytorch-lightning         1.1.8                    pypi_0    pypi
[conda] segmentation-models-pytorch 0.1.3                    pypi_0    pypi
[conda] torch                     2.7.1                    pypi_0    pypi
[conda] torchaudio                0.12.1+cu116             pypi_0    pypi
[conda] torchmetrics              1.3.2                    pypi_0    pypi
[conda] torchvision               0.17.2                   pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] triton                    3.3.1                    pypi_0    pypi
```

**Comment by Alex17Li** [comment]
So far the only way I have found to fix every case I tested (loading just scheduler, just optimizer, both, and changing the lr when just loading optimizer) is to load the optimizer state dict before initializing the scheduler (which is wrong but helps setup the scheduler correctly) and then playing with the optimizer internals

```
base_params=optimizer.param_groups
optimizer.load_state_dict(ckpt_data["optimizer_states"][0])

for opt_param, set_param in zip(optimizer.param_groups, base_params):
    # In case the inintial lr has changed from the last time we loaded optimzer states
    opt_param["initial_lr"] = set_param["lr"]

```

then (in the case of loading a scheduler state dict as well) after constructing the scheduler running this block to play with the optimizer internals again

```
for opt_param, last_lr in zip(optimizer.param_groups, scheduler_state_dict_to_load["_last_lr"]):
     opt_param["lr"] = last_lr
```

**Comment by janeyx99** [comment]
I will not have time to look into this until 2 weeks from now but from a glance I agree this is confusing. If anyone else gets to investigating this before I do, please feel free to open a PR/explain here!

**Comment by krastogi-in** [comment]
@Alex17Li , We might not need this `opt.load_state_dict(state_dict=state['opt'])` as the state variable has old value which was saved before working with new lr . Since old state is getting updated in opt it is expected to have old values and if we do not load the state it should work as expected. Do let me know if  I have wrong assumption

**Comment by Alex17Li** [comment]
Optimizer states are more than just a scalar with 'lr' in them though, you always have to load them, e.g. for some information about the accumulated gradient for momentum methods. So the test case may pass if you remove that line but the assumption is that you want to load optimizer states
